meshmode/mesh/refinement/tesselate.py

```diff
@@ -23,105 +24,283 @@
+    #
+    # * lines and squares don't require any reordering
+
+    return [child[i] for i in [0, 1, 2, 4][:shape.dim + 1]]
+
+# }}}
+
+
+# {{{ resampling
+
+def get_ref_midpoints(shape, ref_vertices):
+    r"""The reference element is considered to be, e.g. for a 2 simplex::
+
+        F
+        | \
+        |   \
+        D----E
+        |   /| \
+        | /  |   \
+        A----B----C
+
+    where the midpoints are ``(B, E, D)``. The same applies to other shapes
+    and higher dimensions.
+
+    :arg ref_vertices: a :class:`list` of node index :class:`tuple`\ s
+        on :math:`[0, 2]^d`.
+    """
+
+    from pytools import add_tuples
+    space = mp.space_for_shape(shape, 1)
+    orig_vertices = [
+            add_tuples(vt, vt) for vt in mp.node_tuples_for_space(space)
+            ]
+    return [rv for rv in ref_vertices if rv not in orig_vertices]
+
+
+def map_unit_nodes_to_children(shape, unit_nodes, tesselation):
+    """
+    :arg unit_nodes: an :class:`~numpy.ndarray` of shape ``(dim, nnodes)``.
+    :arg tesselation: a :class:`TesselationInfo`.
+    """
+
+    ref_vertices = np.array(tesselation.ref_vertices, dtype=np.float).T
+    assert len(unit_nodes.shape) == 2
+
+    for child_element in tesselation.children:
+        indices = get_child_basis_vertex_indices(shape, child_element)
+
+        origin = ref_vertices[:, indices[0]].reshape(-1, 1)
+        basis = ref_vertices[:, indices[1:]] - origin
+
+        # mapped nodes are on [0, 2], so we subtract 1 to get it to [-1, 1]
+        yield basis.dot((unit_nodes + 1.0) / 2.0) + origin - 1.0
+
+
+def get_group_midpoints(group, tesselation, elements):
+    """Compute the midpoints of the vertices of the specified elements.
+
+    :arg group: an instance of :class:`meshmode.mesh.MeshElementGroup`.
+    :arg tesselation: a :class:`TesselationInfo`.
+    :arg elements: a list of (group-relative) element numbers.
+
+    :return: A :class:`dict` mapping element numbers to midpoint
+        coordinates, with each value in the map having shape
+        ``(ambient_dim, nmidpoints)``. The ordering of the midpoints
+        follows their ordering in the tesselation.
+    """
+    from meshmode.mesh import _ModepyElementGroup
+    if not isinstance(group, _ModepyElementGroup):
+        raise TypeError(f"groups of type '{type(group.mesh_el_group).__name__}'"
+                " are not supported.")
+
+    shape = group._modepy_shape
+    space = mp.space_for_shape(shape, group.order)
+
+    # get midpoints in reference coordinates
+    midpoints = -1 + np.array(get_ref_midpoints(shape, tesselation.ref_vertices))
+
+    # resample midpoints to ambient coordinates
+    resampling_mat = mp.resampling_matrix(
+            mp.basis_for_space(space, shape).functions,
+            midpoints.T,
+            group.unit_nodes)
+
+    resampled_midpoints = np.einsum("mu,deu->edm",
+            resampling_mat, group.nodes[:, elements])
+
+    return dict(zip(elements, resampled_midpoints))
```

Comment:
Is it possible to keep/make this an array? `dicts` are expensive. They require Python code to access.

Reply:
Not sure. I'll have to go see how it is used.

Reply:
Hm.. quite a few things in both `Refiner` and `RefinerWithoutAdjacency` want this to be a dict, to match parent/child indices and probably some other things.

Future PR? This was just copied from [here](https://github.com/inducer/meshmode/blob/master/meshmode/mesh/refinement/resampler.py#L91)